app/forms/CommonBinders.scala

```diff
@@ -39,6 +83,172 @@ trait CommonBinders {
+        day   <- stringToIntFormatter("dayEmpty").bind("psoDay", data).right
+        month <- stringToIntFormatter("monthEmpty").bind("psoMonth", data).right
+        year  <- stringToIntFormatter("yearEmpty").bind("psoYear", data).right
+      } yield {
+        (day, month, year)
+      }
+
+      val returnValue = groupedIntsWithCustomErrors fold (
+        errs => stringToIntFormatter("dayEmpty").bind("psoDay", data),
+        success => {
+          val (day, month, year) = (success._1, success._2, success._3)
+          validateCompleteDate(key,Constants.minIP16PSODate,"IP16PsoDetails",day,month,year)
+        }
+      )
+      returnValue
+    }
+    override def unbind(key: String, value: Int) = Map(key -> value.toString)
+  }
+  def dateFormatterFromString: Mapping[Int] = Forms.of[Int](dateStringToIntFormatter)
+
+
+
+  private def dateStringToOptionalIntFormatter = new Formatter[Option[Int]] {
+    override def bind(key: String, data: Map[String, String]) = {
+      val groupedIntsWithCustomErrors: Either[Seq[FormError], (Int, Int, Int, String)] = for {
+        day <- stringToOptionalIntFormatter("dayEmpty").bind("psoDay", data).right
+        month <- stringToOptionalIntFormatter("monthEmpty").bind("psoMonth", data).right
+        year <- stringToOptionalIntFormatter("yearEmpty").bind("psoYear", data).right
+        protectionType <- protectionFormatter.bind("protectionType", data).right
+      } yield {
+        (day.get, month.get, year.get, protectionType)
+      }
+
+      val returnValue: Either[Seq[FormError], Option[Int]] = groupedIntsWithCustomErrors fold(
+        errs => stringToOptionalIntFormatter("dayEmpty").bind("psoDay", data),
+        success => {
+          val (day, month, year, protectionType) = (success._1, success._2, success._3, success._4)
+          val protectionTypeErrorMsg = if (protectionType.toLowerCase.equals("ip2016")) "IP16PsoDetails" else "IP14PsoDetails"
+          val dateComparator = if (protectionType.toLowerCase.equals("ip2016")) Constants.minIP16PSODate else Constants.minIP14PSODate
+          validateCompleteDate(key,dateComparator,protectionTypeErrorMsg,day,month,year).fold[Either[Seq[FormError], Option[Int]]] (
+            errors => Left(errors), value => Right(Some(value))
+          )
+        }
+      )
+      println("THE RETURN ISS "+returnValue)
+      returnValue
+    }
+    override def unbind(key: String, value: Option[Int]) = Map(key -> value.map(_.toString).getOrElse(""))
+
+  }
+  def dateFormatterFromInt: Mapping[Option[Int]] = Forms.of[Option[Int]](dateStringToOptionalIntFormatter)
+
+
+
+  private def protectionFormatter = new Formatter[String] {
+    override def bind(key: String, data: Map[String, String]): Either[Seq[FormError], String] = {
+      data.getOrElse(key, "") match {
+        case success if success.equals("ip2016") || success.equals("ip2014") => Right(success)
+        case other => Left(Seq(FormError(key,"error.required",Nil)))
```

Comment:
This protectionType is passed in as a hidden form parameter so we should throw an exception if it isn't completed.